internal/controller/nonadminbackup_controller.go

```diff
@@ -97,11 +144,279 @@ func (r *NonAdminBackupReconciler) Reconcile(ctx context.Context, req ctrl.Reque
+//
+// Parameters:
+//   - ctx: Context to manage request lifetime.
+//   - logger: Logger instance for logging messages.
+//   - nab: Pointer to the NonAdminBackup object to be managed.
+//
+// This function first checks if the NAB object has the finalizer. If yes, it attempts to locate the associated
+// VeleroBackup object by the UUID stored in the NAB status. If a unique VeleroBackup object is found,
+// deletion is initiated, and the reconcile loop will be requeued. If multiple VeleroBackup objects
+// are found with the same label, the function logs an error and returns.
+//
+// Returns:
+//   - A boolean indicating whether to requeue the reconcile loop,
+//   - An error if VeleroBackup deletion or retrieval fails, for example, due to multiple VeleroBackup objects found with the same UUID label.
+func (r *NonAdminBackupReconciler) createVeleroDeleteBackupRequest(ctx context.Context, logger logr.Logger, nab *nacv1alpha1.NonAdminBackup) (bool, error) {
+	if !controllerutil.ContainsFinalizer(nab, constant.NabFinalizerName) ||
+		nab.Status.VeleroBackup == nil ||
+		nab.Status.VeleroBackup.NameUUID == constant.EmptyString {
+		return false, nil
+	}
+
+	// Initiate deletion of the VeleroBackup object only when the finalizer exists.
+	// For the ForceDelete we do not create DeleteBackupRequest
+	veleroBackupNameUUID := nab.Status.VeleroBackup.NameUUID
+	veleroBackup, err := function.GetVeleroBackupByLabel(ctx, r.Client, r.OADPNamespace, veleroBackupNameUUID)
+
+	if err != nil {
+		// Log error if multiple VeleroBackup objects are found
+		logger.Error(err, findSingleVBError, uuidString, veleroBackupNameUUID)
+		return false, err
+	}
+
+	if veleroBackup == nil {
+		logger.V(1).Info("VeleroBackup already deleted", "UUID", veleroBackupNameUUID)
+		return false, nil
+	}
+
+	deleteBackupRequest, err := function.GetVeleroDeleteBackupRequestByLabel(ctx, r.Client, r.OADPNamespace, veleroBackupNameUUID)
+	if err != nil {
+		// Log error if multiple DeleteBackupRequest objects are found
+		logger.Error(err, findSingleVDBRError, uuidString, veleroBackupNameUUID)
+		return false, err
+	}
+
+	if deleteBackupRequest == nil {
+		// Build the delete request for VeleroBackup created by NAC
+		deleteBackupRequest = builder.ForDeleteBackupRequest(r.OADPNamespace, "").
+			BackupName(veleroBackup.Name).
+			ObjectMeta(
+				builder.WithLabels(
+					velerov1.BackupNameLabel, label.GetValidName(veleroBackup.Name),
+					velerov1.BackupUIDLabel, string(veleroBackup.UID),
+					constant.NabOriginNameUUIDLabel, veleroBackupNameUUID,
+				),
+				builder.WithLabelsMap(function.GetNonAdminLabels()),
+				builder.WithAnnotationsMap(function.GetNonAdminBackupAnnotations(nab.ObjectMeta)),
+				builder.WithGenerateName(veleroBackup.Name+"-"),
+			).Result()
+
+		// Use CreateRetryGenerateName for retry logic in creating the delete request
+		if err := function.CreateRetryGenerateName(r.Client, ctx, deleteBackupRequest); err != nil {
```

Comment:
I think we should not create function for this

Using this https://github.com/migtools/oadp-non-admin/issues/106 would solve the problem here, no?

Reply:
It's exactly how Velero CLI is doing, so I want this part to be very similar. If you go to the function.CreateRetryGenerateName you will see comment pointing to the Velero part from which the logic was borrowed.

Reply:
I think this retry is not needed (if create fails, reconcile exits with error and then it will enter the reconcile again and retry creation here)

but if you want to use this function, then delete duplication from `internal/common/function/function.go`, just import from velero package directly

Reply:
ok will use that from velero directly.

Reply:
yeah we could directly re-use from velero.

Reply:
Used directly from velero in the newest commit.